cmd/compute-domain-kubelet-plugin/device_state.go

```diff
@@ -153,50 +158,67 @@ func (s *DeviceState) Prepare(ctx context.Context, claim *resourceapi.ResourceCl
-		return nil, fmt.Errorf("unable to sync to checkpoint: %v", err)
+	// Add ResourceClaimStatus API object to node-local checkpoint: the
+	// 'unprepare' code path must use local state exclusively (ResourceClaim
+	// object might have been deleted from the API server).
+	preparedClaimStates[claimUID] = claim.Status
+	preparedDevicesByClaimUID[claimUID] = preparedDevices
+	if err := s.checkpointManager.CreateCheckpoint(DriverPluginCheckpointFileBasename, checkpoint); err != nil {
+		return nil, fmt.Errorf("unable to create checkpoint: %w", err)
 	}
+	klog.V(6).Infof("checkpoint written for claim %v", claimUID)
 
-	return preparedClaims[claimUID].GetDevices(), nil
+	return preparedDevicesByClaimUID[claimUID].GetDevices(), nil
 }
 
-func (s *DeviceState) Unprepare(ctx context.Context, claim *resourceapi.ResourceClaim) error {
+func (s *DeviceState) Unprepare(ctx context.Context, claimRef kubeletplugin.NamespacedObject) error {
 	s.Lock()
 	defer s.Unlock()
 
-	claimUID := string(claim.UID)
-
-	if err := s.unprepareDevices(ctx, claim); err != nil {
-		return fmt.Errorf("unprepare devices failed: %w", err)
-	}
+	claimUID := string(claimRef.UID)
 
+	// Rely on local checkpoint state for ability to clean up.
 	checkpoint := newCheckpoint()
-	if err := s.checkpointManager.GetCheckpoint(DriverPluginCheckpointFile, checkpoint); err != nil {
-		return fmt.Errorf("unable to sync from checkpoint: %v", err)
+	if err := s.checkpointManager.GetCheckpoint(DriverPluginCheckpointFileBasename, checkpoint); err != nil {
+		return fmt.Errorf("unable to get checkpoint: %w", err)
+	}
+
+	preparedClaimStates := checkpoint.V1.PreparedClaimStates
+	status, found := preparedClaimStates[claimUID]
+	if !found {
+		// Not an error: if this claim UID is not in the checkpoint then this
+		// device was never prepared (assume that Prepare+Checkpoint are done
+		// transactionally). Note that claimRef.String() contains namespace,
+		// name, UID.
+		klog.Infof("unprepare noop: claim not found in checkpoint data: %v", claimRef.String())
+		return nil
 	}
-	preparedClaims := checkpoint.V1.PreparedClaims
 
-	if preparedClaims[claimUID] == nil {
-		return nil
+	if err := s.unprepareDevices(ctx, &status); err != nil {
+		return fmt.Errorf("unprepare devices failed: %w", err)
 	}
 
 	err := s.cdi.DeleteClaimSpecFile(claimUID)
 	if err != nil {
 		return fmt.Errorf("unable to delete CDI spec file for claim: %w", err)
 	}
 
-	delete(preparedClaims, claimUID)
-	if err := s.checkpointManager.CreateCheckpoint(DriverPluginCheckpointFile, checkpoint); err != nil {
-		return fmt.Errorf("unable to sync to checkpoint: %v", err)
+	// Write new checkpoint reflecting that all devices for this claim have been
+	// unprepared (by virtue of removing its UID from all mappings).
+	preparedDevicesByClaimUID := checkpoint.V1.PreparedDevicesByClaimUID
+	delete(preparedDevicesByClaimUID, claimUID)
+	delete(preparedClaimStates, claimUID)
```

Comment:
maybe instead have just one top-level map, and every value in that map would be an object with a `.Status` and a `.PreparedDevices` property.

Reply:
done